File: packages/node-agent/src/processes/identity.ts
import { createHash } from 'node:crypto';
import path from 'node:path';
import type { JsonObject, SandboxConfig } from '../types.js';
import { normalizedSandboxConfig } from '../sandbox.js';

export interface ProcessCommandIdentitySpec {
  program: string;
  argv: string[];
  display: string;
  shell: boolean;
}

function stableValue(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(stableValue);
  if (value && typeof value === 'object') {
    return Object.fromEntries(Object.entries(value as Record<string, unknown>)
      .sort(([left], [right]) => left.localeCompare(right))
      .map(([key, nested]) => [key, stableValue(nested)]));
  }
  return value;
}

function sandboxFingerprintMaterial(config: SandboxConfig | undefined): unknown {
  const normalized = normalizedSandboxConfig(config);
  if (!normalized.enabled) return null;
  return stableValue({
    backend: normalized.backend,
    external_paths: normalized.externalPaths,
    options: normalized.options
  });
}

export function commandFingerprint(
  cwd: string,
  spec: ProcessCommandIdentitySpec,
  args: JsonObject,
  timeoutMs: number,
  sandboxConfig?: SandboxConfig
): string {
  const env = Object.fromEntries(Object.entries((args.env as Record<string, unknown> | undefined) ?? {})
    .sort(([left], [right]) => left.localeCompare(right))
    .map(([name, value]) => [name, String(value)]));
  const removeEnv = Array.isArray(args.remove_env) ? args.remove_env.map(String).sort() : [];
  const stdin = typeof args.stdin === 'string' ? args.stdin : '';
  const material = stableValue({
    cwd,
    program: spec.program,
    argv: spec.argv,
    shell: spec.shell,
    env,
    remove_env: removeEnv,
    timeout_ms: timeoutMs,
    tty: args.tty === true,
    stdin_sha256: createHash('sha256').update(stdin).digest('hex'),
    post_checks: Array.isArray(args.post_checks) ? args.post_checks.slice(0, 16) : [],
    resource_lock_group: String(args.lock_group ?? '').trim() || null,
    sandbox: sandboxFingerprintMaterial(sandboxConfig)
  });
  return createHash('sha256').update(JSON.stringify(material)).digest('hex');
}

export function safeAutomaticDedup(spec: ProcessCommandIdentitySpec): boolean {
  const executable = (spec.program.split(/[\\/]/).at(-1) ?? '').toLowerCase().replace(/\.exe$/, '').replace(/\.cmd$/, '');
  if (executable !== 'cargo') return false;
  return ['check', 'test', 'build', 'fmt', 'clippy'].includes(spec.argv[0] ?? '');
}

function commandArgumentValue(argv: string[], name: string): string | undefined {
  for (let index = 0; index < argv.length; index += 1) {
    const argument = argv[index] ?? '';
    if (argument === name) return argv[index + 1];
    if (argument.startsWith(`${name}=`)) return argument.slice(name.length + 1);
  }
  return undefined;
}

function isCargoCommand(spec: ProcessCommandIdentitySpec): boolean {
  const executable = (spec.program.split(/[\\/]/).at(-1) ?? '').toLowerCase().replace(/\.exe$/, '').replace(/\.cmd$/, '');
  const display = spec.display.toLowerCase();
  return executable === 'cargo' || display.includes('cargo ') || display.includes('tauri build');
}

export function cargoTargetLock(
  cwd: string,
  spec: ProcessCommandIdentitySpec,
  args: JsonObject
): { group: string; target: string } | undefined {
  if (!isCargoCommand(spec)) return undefined;
  const env = (args.env as Record<string, unknown> | undefined) ?? {};
  const envTarget = Object.entries(env).find(([key]) => key.toLowerCase() === 'cargo_target_dir')?.[1];
  const configuredTarget = commandArgumentValue(spec.argv, '--target-dir') ?? (envTarget === undefined ? undefined : String(envTarget));
  let target: string;
  if (configuredTarget) {
    target = path.isAbsolute(configuredTarget) ? configuredTarget : path.join(cwd, configuredTarget);
  } else {
    const manifest = commandArgumentValue(spec.argv, '--manifest-path');
    if (manifest) {
      const manifestPath = path.isAbsolute(manifest) ? manifest : path.join(cwd, manifest);
      target = path.join(path.dirname(manifestPath), 'target');
    } else if (spec.display.toLowerCase().includes('tauri') && path.basename(cwd).toLowerCase() !== 'src-tauri') {
      target = path.join(cwd, 'src-tauri', 'target');
    } else {
      target = path.join(cwd, 'target');
    }
  }
  const normalizedTarget = path.resolve(target);
  const digest = createHash('sha256').update(normalizedTarget).digest('hex');
  return { group: `cargo-target:${digest.slice(0, 24)}`, target: normalizedTarget };
}

File: src-tauri/src/tools/exec/identity.rs
use std::path::{Path, PathBuf};

use serde_json::{json, Value};
use sha2::{Digest, Sha256};

use crate::mcp::command_kind;

use super::spec::{ExecSpec, PostCheckSpec};

#[derive(Clone, Debug)]
pub(super) struct ExecutionIdentity {
    pub(super) operation_id: Option<String>,
    pub(super) command_fingerprint: String,
    pub(super) resource_lock_group: Option<String>,
    pub(super) resource_lock_target: Option<String>,
}

pub(super) fn sha256_hex(bytes: &[u8]) -> String {
    format!("{:x}", Sha256::digest(bytes))
}

fn is_cargo_command(spec: &ExecSpec) -> bool {
    Path::new(&spec.program)
        .file_stem()
        .and_then(|value| value.to_str())
        .is_some_and(|value| value.eq_ignore_ascii_case("cargo"))
        || spec.display.to_ascii_lowercase().contains("cargo ")
        || spec.display.to_ascii_lowercase().contains("tauri build")
}

fn command_argument_value(args: &[String], name: &str) -> Option<String> {
    for (index, argument) in args.iter().enumerate() {
        if argument == name {
            return args.get(index + 1).cloned();
        }
        if let Some(value) = argument.strip_prefix(&format!("{name}=")) {
            return Some(value.to_string());
        }
    }
    None
}

fn normalized_lock_path(path: PathBuf) -> PathBuf {
    std::fs::canonicalize(&path).unwrap_or(path)
}

pub(super) fn cargo_target_lock(spec: &ExecSpec, cwd: &Path) -> Option<(String, String)> {
    if !is_cargo_command(spec) {
        return None;
    }
    let env_target = spec
        .env
        .iter()
        .find(|(key, _)| key.eq_ignore_ascii_case("CARGO_TARGET_DIR"))
        .map(|(_, value)| value.clone());
    let target = command_argument_value(&spec.args, "--target-dir")
        .or(env_target)
        .map(PathBuf::from)
        .map(|path| {
            if path.is_absolute() {
                path
            } else {
                cwd.join(path)
            }
        })
        .or_else(|| {
            command_argument_value(&spec.args, "--manifest-path").map(|manifest| {
                let manifest = PathBuf::from(manifest);
                let manifest = if manifest.is_absolute() {
                    manifest
                } else {
                    cwd.join(manifest)
                };
                manifest.parent().unwrap_or(cwd).join("target")
            })
        })
        .or_else(|| {
            let lower = spec.display.to_ascii_lowercase();
            let tauri_root = cwd.join("src-tauri");
            (lower.contains("tauri") && tauri_root.join("Cargo.toml").is_file())
                .then(|| tauri_root.join("target"))
        })
        .unwrap_or_else(|| cwd.join("target"));
    let target = normalized_lock_path(target);
    let display = target.to_string_lossy().into_owned();
    let digest = sha256_hex(display.as_bytes());
    Some((format!("cargo-target:{}", &digest[..24]), display))
}

pub(super) fn execution_identity(
    args: &Value,
    spec: &ExecSpec,
    cwd: &Path,
    timeout_ms: u64,
    tty: bool,
    stdin_text: &str,
    post_checks: &[PostCheckSpec],
) -> ExecutionIdentity {
    let mut env = spec.env.clone();
    env.sort();
    let mut remove_env = spec.remove_env.clone();
    remove_env.sort();
    let post_checks = post_checks
        .iter()
        .map(|check| {
            json!({
                "name": check.name,
                "program": check.exec.program,
                "args": check.exec.args,
                "shell": check.exec.shell,
                "env": check.exec.env,
                "remove_env": check.exec.remove_env,
                "expected_exit_code": check.expected_exit_code,
                "timeout_ms": check.timeout.as_millis(),
                "max_output_bytes": check.max_output_bytes
            })
        })
        .collect::<Vec<_>>();
    let automatic_cargo_dedupe = is_cargo_command(spec)
        && matches!(
            command_kind(args),
            "cargo_test" | "cargo_check" | "build" | "format"
        );
    let explicit_operation_id = args
        .get("operation_id")
        .and_then(Value::as_str)
        .filter(|value| !value.trim().is_empty())
        .map(str::to_string);
    let deduplicate = explicit_operation_id.is_some()
        || args
            .get("deduplicate")
            .and_then(Value::as_bool)
            .unwrap_or(automatic_cargo_dedupe);
    let automatic_lock = cargo_target_lock(spec, cwd);
    let resource_lock_group = args
        .get("lock_group")
        .and_then(Value::as_str)
        .filter(|value| !value.trim().is_empty())
        .map(str::to_string)
        .or_else(|| automatic_lock.as_ref().map(|(group, _)| group.clone()));
    let resource_lock_target = automatic_lock.map(|(_, target)| target);
    let material = json!({
        "cwd": cwd.to_string_lossy(),
        "program": spec.program,
        "args": spec.args,
        "shell": spec.shell,
        "env": env,
        "remove_env": remove_env,
        "timeout_ms": timeout_ms,
        "tty": tty,
        "stdin_sha256": sha256_hex(stdin_text.as_bytes()),
        "post_checks": post_checks,
        "resource_lock_group": resource_lock_group
    });
    let command_fingerprint = sha256_hex(&serde_json::to_vec(&material).unwrap_or_default());
    let operation_id = explicit_operation_id
        .or_else(|| deduplicate.then(|| format!("auto:{}", &command_fingerprint[..32])));
    ExecutionIdentity {
        operation_id,
        command_fingerprint,
        resource_lock_group,
        resource_lock_target,
    }
}
